Fix Language Wrapping

zeroSteiner opened the pull request:
I suggest reviewing the code without whitespace changes. I converted some tab characters to spaces.

This fixes #51. The `#hexify` and `#to_$lang` methods were not correctly wrapping at the specified number of columns. This also adds a unit test to show that the wrapping is working now, and to ensure that it continues to work as expected in the future.

## Before

```
[10] pry(main)> puts Rex::Text.to_ruby("A"*50, 60)
buf = 
"\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41" +
"\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41" +
"\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41" +
"\x41\x41\x41\x41\x41\x41\x41\x41"
=> nil
[11] pry(main)> puts Rex::Text.to_python("A"*50, 60)
buf =  b""
buf += b"\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41"
buf += b"\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41"
buf += b"\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41"
buf += b"\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41"
=> nil
[12] pry(main)> puts Rex::Text.to_csharp("A"*50, 60)
byte[] buf = new byte[50] {
0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,
0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,
0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,
0x41,0x41,0x41,0x41,0x41 };
=> nil
[13] pry(main)> puts Rex::Text.to_golang("A"*50, 60)
 buf :=  []byte{
0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 
0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 
0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 
0x41, 0x41, 0x41, 0x41, 0x41 }
=> nil
[14] pry(main)> 
```

## After 

```
[1] pry(main)> require 'rex/text'
=> true
[2] pry(main)> puts Rex::Text.to_ruby("A"*50, 60)
buf = 
"\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41" +
"\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41" +
"\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41" +
"\x41\x41\x41\x41\x41\x41\x41\x41"
=> nil
[3] pry(main)> puts Rex::Text.to_python("A"*50, 60)
buf =  b""
buf += b"\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41"
buf += b"\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41"
buf += b"\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41"
buf += b"\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41\x41"
buf += b"\x41\x41"
=> nil
[4] pry(main)> puts Rex::Text.to_csharp("A"*50, 60)
byte[] buf = new byte[50] {0x41,0x41,0x41,0x41,0x41,0x41,
0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,
0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,
0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,
0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41};
=> nil
[5] pry(main)> puts Rex::Text.to_golang("A"*50, 60)
buf :=  []byte{0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,
0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,
0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,
0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,0x41,
0x41,0x41,0x41,0x41,0x41};
=> nil
[6] pry(main)>
```

adfoster-r7 commented on spec/rex/text/hex_spec.rb:
```diff
@@ -0,0 +1,23 @@
+
+RSpec.describe Rex::Text do
+  context "Class Hex methods" do
+    context ".hexify" do
```
```suggestion
    describe ".hexify" do
```

adfoster-r7 commented on spec/rex/text/hex_spec.rb:
```diff
@@ -0,0 +1,23 @@
+require 'spec_helper'
+
+RSpec.describe Rex::Text do
+  context "Class Hex methods" do
```
Can we nuke the outer `context` here, it's redundant with the `describe '.hexify' do` below :+1:

```suggestion
```

adfoster-r7 replied:
Ah, this is the existing convention in the codebase. Hm. I'd be good to nuke it still 👀

adfoster-r7 commented on spec/rex/text/lang_spec.rb:
```diff
@@ -0,0 +1,21 @@
+require 'spec_helper'
+
+RSpec.describe Rex::Text do
+  context "Class Language methods" do
```
```suggestion
```

adfoster-r7 commented on spec/rex/text/lang_spec.rb:
```diff
@@ -0,0 +1,21 @@
+
+    LANGUAGES = %w[ bash c csharp golang perl python ruby ]
+
+    context "to language methods should wrap at the specified columns" do
+      LANGUAGES.each do |lang|
+        context ".to_#{lang}" do
```
```suggestion
    LANGUAGES.each do |lang|
      describe ".to_#{lang}" do
```

adfoster-r7 commented on spec/rex/text/hex_spec.rb:
```diff
@@ -0,0 +1,23 @@
+      let(:wrap) { 60 }
+      it 'should wrap at the specified columns' do
+        lines = described_class.hexify("A" * 100, wrap).split("\n")
+        expect(lines.any? { |line| line.rstrip.length > wrap }).to be_falsey
```
not a blocker: We've got similar matchers for the wrapped table functionality that could be used:

https://github.com/rapid7/rex-text/blob/4bfa153b34a4c1f2086ad990b57c7bf2f861f54b/spec/rex/text/wrapped_table_spec.rb#L6-L13

https://github.com/rapid7/rex-text/blob/4bfa153b34a4c1f2086ad990b57c7bf2f861f54b/spec/rex/text/wrapped_table_spec.rb#L111

adfoster-r7 replied:
Which is nice, as iirc it'll log which row failed to have the correct amount of characters - whilst this current implementation would fail with a true/false, and the developer would have to put breakpoints in to see what's gone wrong 🕵️

adfoster-r7 commented:
Not a blocker, just noticed it when playing around with values:

```
2.7.2 :023 > puts Rex::Text.to_ruby("A"* 1, 5)
buf = 
"" +
"\x41
"
```

adfoster-r7 reviewed (changes requested):
Just one blocker for me, the scenario of generating invalid code for empty buffers:

https://github.com/rapid7/rex-text/pull/53#issuecomment-1233351951

zeroSteiner commented:
- [x] Fixed golang / csharp buffer generation for empty buffers
- [x] Implemented all the feedback on the specs, moved the `have_maximum_width` to `spec_helper` for reuse
- [x] Started to raise an exception in `#hexify` when the line start, line end won't fit with a single character because the width is too small
- [ ] **Did not** fix java. It actually [does not](https://github.com/rapid7/rex-text/blob/b6cb69c5a6d4cd74af303f6d270652f6b1739d1e/lib/rex/text/lang.rb#L101) support wrapping right now. The second argument is the name not wrap. Adding wrapping support would make it the only one where name and wrap are swapped with each other which is disappointing.
